### DISCUSSION TITLE
IDocumentSession.Query<> returns null when searching for a single document

### DISCUSSION THREAD

**Discussion by johnholliday** [JasperFx/marten#2718] [General]
The following code works but requires loading the entire collection.

```csharp
var templates = await _session.Query<Template>().ToListAsync();
if (templates != nuLL) {
  foreach (var item in templates)
    if (item.Slug == slug) return item;
}
return default;
```
I have a potentially very large collection of documents, so I need to retrieve the specific documents I'm looking for.  However, the following code always returns null because the sequence contains no elements.

```csharp
var template = await _session.Query<Template>().Where(x => x.Slug == slug).SingleOrDefaultAsync();
if (template != null) return template;
return default;
```
This also fails similarly.
```csharp
var template = await _session.Query<Template>().SingleOrDefaultAsync(x => x.Slug == slug);
if (template != null) return template;
return default;
```

**Comment by mysticmind** [comment]
Query for a single document should work fine. What is the value of slug which you are searching and the entry in db? It would help for you to provide some code with data for us to check.

**Comment by johnholliday** [comment]
This is the method:

```csharp
    public async Task<Template?> GetBySlugAsync(string slug)
    {
        var templates = await _session.Query<Template>().ToListAsync();
        if (templates != null) foreach (var item in templates) if (item.Slug == slug) return item;
        // return await _session.Query<Template>().Where(x => x.Slug == slug).SingleOrDefaultAsync();
        return default;
    }
```

This is value of "slug" that is being passed to the method:

```csharp
var slug = "my-document-template";
```

This is the entry in the database:

```json
{
  "id": "57b5239b-171d-4200-9bd2-ed7951d76e72",
  "title": "My Document Template",
  "slug": "my-document-template",
  "description": "a brief Document description",
  "tags": [
    "first tag",
    "another tag"
  ]
}
```
Retrieving the entire list and then matching on the slug returns the correct database entry.
Querying for a single document returns null.

**Comment by jeremydmiller** [comment]
@johnholliday Hey, check your JSON settings in Marten. Notice that "slug" is I'm guessing camel casing in your database. Are you explicitly setting your JSON casing to CamelCase with Marten?

I'm moving this to a discussion, because it's extremely unlikely that this is anything but the JSON serialization settings.
